docs/ShapeInference.md

```diff
@@ -52,22 +87,51 @@ To see numerous examples, search for occurrences of
 relatively involved is the implementation for `Concat`, in
 onnx/defs/tensor/defs.cc.
 
-## Limitations
-
-Shape inference is not guaranteed to be complete. In particular, some
-dynamic behaviors block the flow of shape inference, for example a
-Reshape to a dynamically-provide shape. Also, all operators are not
-required to have a shape inference implementation.
-
-Shape inference works only with constants and simple variables. It
-does not support arithmetic expressions containing variables. For
-example, `Concat` on tensors of shapes `(5, 2)` and `(7, 2)` can be
-inferred to produce a result of shape `(12, 2)`, but `Concat` on
-tensors of shapes `(5, 2)` and `(N, 2)` will simply produce `(M, 2)`,
-rather than containing a representation of `N+5`. Note that differing
-unknown symbolic values will be propagated, so the `M` here represents
-an unknown quantity that is the same as other occurrences of `M`.
+Please note the following points when implementing the shape-inference method for
+operators to avoid common errors:
+
+* Before accessing the `shape` of any input, the code must check that
+the shape is available. If unavailable, it should be treated as a dynamic
+tensor whose rank is unknown and handled appropriately. Usually, the
+shape-inference logic is guarded by a call to `hasInputShape` or
+`hasNInputShapes`.
+
+* Before accessing the `dim_value` or `dim_param` of any dimension, the
+code must check if these fields have a value. In particular, the code must
+handle the possibility that the dimension may not have a statically
+known value.
+
+There are several utility functions in [shape_inference.h](/onnx/defs/shape_inference.h)
```

Comment:
Should we make sure users are using the utility functions to throw exceptions if one shape has an unexpecgted value?

Reply:
Yes, but, unfortunately, we are dependent on reviewers to check for this. Any other suggestions?